File: bridge/src/pixoo/micro-glyphs.ts
/**
 * Timebox Mini 11×11 "Agent Beacon" renderer.
 *
 * The panel is too small for a miniature terrarium. It gets its own visual
 * language instead: a stable 9×9 official agent mark sits inside a one-pixel
 * status rail. Identity never deforms; only the rail animates.
 *
 * The 9×9 masks are generated directly from design/brand/*.svg. This file owns
 * only device-specific color and motion, keeping geometry canonical while
 * making every one of the Timebox's 121 LEDs intentional.
 */

import {
  OFFICIAL_TIMEBOX_GLYPHS,
  OFFICIAL_TIMEBOX_GLYPH_SIZE,
  type OfficialDotGlyphName,
} from './official-dot-glyphs.generated.js';

export type RGB = readonly [number, number, number];
export type MicroCreature = 'octopus' | 'jellyfish' | 'opencode' | 'crayfish' | 'antigravity' | 'kiro';
export type MicroAggregate = 'idle' | 'processing' | 'awaiting' | 'error';

export const MICRO_SIZE = 11;

const BACKGROUND: RGB = [2, 6, 10];
const IDLE_RAIL: RGB = [38, 170, 116];
const PROCESSING_RAIL: RGB = [82, 220, 255];
const AWAITING_RAIL: RGB = [255, 184, 54];
const ERROR_RAIL: RGB = [255, 70, 70];

const AGENT_COLORS: Record<Exclude<OfficialDotGlyphName, 'antigravity'>, RGB> = {
  claudeCode: [235, 130, 90],
  codex: [112, 124, 255],
  openCode: [238, 238, 238],
  openClaw: [255, 92, 92],
  kiro: [124, 58, 237],
};

const PERIMETER: ReadonlyArray<readonly [number, number]> = [
  ...Array.from({ length: 11 }, (_, x) => [x, 0] as const),
  ...Array.from({ length: 10 }, (_, i) => [10, i + 1] as const),
  ...Array.from({ length: 10 }, (_, i) => [9 - i, 10] as const),
  ...Array.from({ length: 9 }, (_, i) => [0, 9 - i] as const),
];

function setPixel(buf: Uint8Array, x: number, y: number, color: RGB, intensity = 1): void {
  if (x < 0 || x >= MICRO_SIZE || y < 0 || y >= MICRO_SIZE) return;
  const i = (y * MICRO_SIZE + x) * 3;
  buf[i] = Math.round(color[0] * intensity);
  buf[i + 1] = Math.round(color[1] * intensity);
  buf[i + 2] = Math.round(color[2] * intensity);
}

function officialName(creature: MicroCreature): OfficialDotGlyphName {
  switch (creature) {
    case 'octopus': return 'claudeCode';
    case 'jellyfish': return 'codex';
    case 'opencode': return 'openCode';
    case 'crayfish': return 'openClaw';
    case 'antigravity': return 'antigravity';
    case 'kiro': return 'kiro';
  }
}

function antigravityColor(sourceX: number): RGB {
  const bands: RGB[] = [
    [92, 214, 77], [245, 203, 36], [255, 132, 16],
    [255, 82, 65], [183, 92, 182], [102, 111, 225], [36, 126, 255],
  ];
  const index = Math.min(bands.length - 1, Math.floor(sourceX * bands.length / OFFICIAL_TIMEBOX_GLYPH_SIZE));
  return bands[index];
}

function paintOfficialMark(buf: Uint8Array, creature: MicroCreature, aggregate: MicroAggregate): void {
  const name = officialName(creature);
  const mask = OFFICIAL_TIMEBOX_GLYPHS[name];
  const stateIntensity = aggregate === 'idle' ? 0.92 : aggregate === 'error' ? 0.72 : 1;

  for (let y = 0; y < OFFICIAL_TIMEBOX_GLYPH_SIZE; y++) {
    for (let x = 0; x < OFFICIAL_TIMEBOX_GLYPH_SIZE; x++) {
      const alpha = mask[y * OFFICIAL_TIMEBOX_GLYPH_SIZE + x];
      // Four deliberate levels survive the Timebox's 4-bit channel packing while
      // restoring the contour/shading that the old two-step body reduced to a
      // flat block. Keep true cutouts black so the official negative space reads.
      const coverage = alpha >= 224 ? 1 : alpha >= 144 ? 0.82 : alpha >= 56 ? 0.56 : alpha >= 20 ? 0.32 : 0;
      if (coverage === 0) continue;
      const color = name === 'antigravity' ? antigravityColor(x) : AGENT_COLORS[name];
      const light = 0.88 + (1 - y / (OFFICIAL_TIMEBOX_GLYPH_SIZE - 1)) * 0.12;
      setPixel(buf, x + 1, y + 1, color, coverage * stateIntensity * light);
    }
  }

  // The official silhouette carries the lobster identity; two cyan eye pixels
  // remain a device-tuned accessibility accent on the 9px reduction.
  if (name === 'openClaw') {
    setPixel(buf, 4, 4, [0, 229, 204], stateIntensity);
    setPixel(buf, 7, 4, [0, 229, 204], stateIntensity);
  }
}

function paintStandby(buf: Uint8Array, animFrame: number): void {
  const pulse = 0.55 + 0.25 * ((Math.sin(animFrame * 0.18) + 1) / 2);
  const tide: RGB = [76, 206, 220];
  for (const [y, left, right] of [[4, 4, 6], [6, 3, 7], [8, 4, 6]] as const) {
    for (let x = left; x <= right; x++) setPixel(buf, x, y, tide, pulse);
  }
}

function paintStatusRail(buf: Uint8Array, aggregate: MicroAggregate, animFrame: number): void {
  const railColor = aggregate === 'processing' ? PROCESSING_RAIL
    : aggregate === 'awaiting' ? AWAITING_RAIL
      : aggregate === 'error' ? ERROR_RAIL : IDLE_RAIL;
  const baseIntensity = aggregate === 'idle' ? 0.10 : 0.13;
  for (const [x, y] of PERIMETER) setPixel(buf, x, y, railColor, baseIntensity);

  switch (aggregate) {
    case 'processing': {
      const head = Math.floor(animFrame / 3) % PERIMETER.length;
      for (let trail = 0; trail < 5; trail++) {
        const [x, y] = PERIMETER[(head - trail + PERIMETER.length) % PERIMETER.length];
        setPixel(buf, x, y, PROCESSING_RAIL, 1 - trail * 0.17);
      }
      break;
    }
    case 'awaiting': {
      const phase = (Math.floor(animFrame / 4) & 1) === 0;
      const points = phase
        ? [[0, 0], [1, 0], [0, 1], [10, 10], [9, 10], [10, 9]]
        : [[10, 0], [9, 0], [10, 1], [0, 10], [1, 10], [0, 9]];
      for (const [x, y] of points) setPixel(buf, x, y, AWAITING_RAIL);
      break;
    }
    case 'error': {
      const intensity = 0.65 + 0.35 * ((Math.sin(animFrame * 0.35) + 1) / 2);
      for (let i = 0; i < PERIMETER.length; i += 2) {
        const [x, y] = PERIMETER[i];
        setPixel(buf, x, y, ERROR_RAIL, intensity);
      }
      break;
    }
    case 'idle': {
      const intensity = 0.56 + 0.16 * ((Math.sin(animFrame * 0.12) + 1) / 2);
      for (const [x, y] of [[0, 0], [10, 0], [10, 10], [0, 10]] as const) {
        setPixel(buf, x, y, IDLE_RAIL, intensity);
      }
      break;
    }
  }
}

/** Paint a complete 11×11 Agent Beacon frame. */
export function paintTimeboxBeacon(
  buf: Uint8Array,
  creature: MicroCreature | null,
  aggregate: MicroAggregate,
  animFrame: number,
): void {
  if (buf.length !== MICRO_SIZE * MICRO_SIZE * 3) return;
  for (let i = 0; i < MICRO_SIZE * MICRO_SIZE; i++) buf.set(BACKGROUND, i * 3);
  if (creature) paintOfficialMark(buf, creature, aggregate);
  else paintStandby(buf, animFrame);
  paintStatusRail(buf, aggregate, animFrame);
}

File: apple/AgentDeck/Daemon/Modules/MicroGlyphs.swift
#if os(macOS)
// MicroGlyphs.swift — Timebox Mini 11×11 "Agent Beacon" renderer.
//
// The Timebox is not a miniature terrarium. A generated 9×9 official agent
// mark stays fixed inside a one-pixel status rail. Only that rail animates, so
// identity remains stable while processing/awaiting/error is legible at a glance.
// Mask geometry comes from design/brand/*.svg via generate-micro-glyphs.mjs;
// this file owns only the Timebox-specific color and motion language.

import Foundation

enum MicroCreature { case octopus, codex, opencode, crayfish, antigravity, kiro }
enum MicroAggregate { case idle, processing, awaiting, error }

enum MicroGlyphs {
    typealias RGB = (UInt8, UInt8, UInt8)
    static let size = 11

    private static let background: RGB = (2, 6, 10)
    private static let idleRail: RGB = (38, 170, 116)
    private static let processingRail: RGB = (82, 220, 255)
    private static let awaitingRail: RGB = (255, 184, 54)
    private static let errorRail: RGB = (255, 70, 70)

    private static let perimeter: [(Int, Int)] = {
        var p: [(Int, Int)] = []
        for x in 0..<11 { p.append((x, 0)) }
        for y in 1..<11 { p.append((10, y)) }
        for x in stride(from: 9, through: 0, by: -1) { p.append((x, 10)) }
        for y in stride(from: 9, through: 1, by: -1) { p.append((0, y)) }
        return p
    }()

    private static func officialGlyph(for creature: MicroCreature) -> OfficialDotGlyph {
        switch creature {
        case .octopus: return .claudeCode
        case .codex: return .codex
        case .opencode: return .openCode
        case .crayfish: return .openClaw
        case .antigravity: return .antigravity
        case .kiro: return .kiro
        }
    }

    private static func agentColor(_ glyph: OfficialDotGlyph, sourceX: Int) -> RGB {
        switch glyph {
        case .claudeCode: return (235, 130, 90)
        case .codex: return (112, 124, 255)
        case .openCode: return (238, 238, 238)
        case .openClaw: return (255, 92, 92)
        case .kiro: return (124, 58, 237)
        case .antigravity:
            let bands: [RGB] = [
                (92, 214, 77), (245, 203, 36), (255, 132, 16),
                (255, 82, 65), (183, 92, 182), (102, 111, 225), (36, 126, 255),
            ]
            let index = min(bands.count - 1, sourceX * bands.count / OfficialTimeboxGlyphs.size)
            return bands[index]
        }
    }

    private static func setPixel(
        _ buf: inout [UInt8], x: Int, y: Int, color: RGB, intensity: Double = 1
    ) {
        guard x >= 0, x < size, y >= 0, y < size else { return }
        func channel(_ value: UInt8) -> UInt8 {
            UInt8(max(0, min(255, Int(round(Double(value) * intensity)))))
        }
        let i = (y * size + x) * 3
        buf[i] = channel(color.0)
        buf[i + 1] = channel(color.1)
        buf[i + 2] = channel(color.2)
    }

    private static func paintOfficialMark(
        _ buf: inout [UInt8], creature: MicroCreature, aggregate: MicroAggregate
    ) {
        let glyph = officialGlyph(for: creature)
        guard let mask = OfficialTimeboxGlyphs.masks[glyph] else { return }
        let stateIntensity = aggregate == .idle ? 0.92 : (aggregate == .error ? 0.72 : 1)
        let n = OfficialTimeboxGlyphs.size

        for y in 0..<n {
            for x in 0..<n {
                let alpha = mask[y * n + x]
                let coverage = alpha >= 224 ? 1.0
                    : (alpha >= 144 ? 0.82 : (alpha >= 56 ? 0.56 : (alpha >= 20 ? 0.32 : 0)))
                guard coverage > 0 else { continue }
                let light = 0.88 + (1 - Double(y) / Double(n - 1)) * 0.12
                setPixel(
                    &buf, x: x + 1, y: y + 1,
                    color: agentColor(glyph, sourceX: x),
                    intensity: coverage * stateIntensity * light
                )
            }
        }

        if glyph == .openClaw {
            setPixel(&buf, x: 4, y: 4, color: (0, 229, 204), intensity: stateIntensity)
            setPixel(&buf, x: 7, y: 4, color: (0, 229, 204), intensity: stateIntensity)
        }
    }

    private static func paintStandby(_ buf: inout [UInt8], animFrame: Int) {
        let pulse = 0.55 + 0.25 * ((sin(Double(animFrame) * 0.18) + 1) / 2)
        let tide: RGB = (76, 206, 220)
        for (y, left, right) in [(4, 4, 6), (6, 3, 7), (8, 4, 6)] {
            for x in left...right { setPixel(&buf, x: x, y: y, color: tide, intensity: pulse) }
        }
    }

    private static func paintStatusRail(
        _ buf: inout [UInt8], aggregate: MicroAggregate, animFrame: Int
    ) {
        let railColor: RGB
        switch aggregate {
        case .processing: railColor = processingRail
        case .awaiting: railColor = awaitingRail
        case .error: railColor = errorRail
        case .idle: railColor = idleRail
        }
        let baseIntensity = aggregate == .idle ? 0.10 : 0.13
        for (x, y) in perimeter {
            setPixel(&buf, x: x, y: y, color: railColor, intensity: baseIntensity)
        }

        switch aggregate {
        case .processing:
            let head = (animFrame / 3) % perimeter.count
            for trail in 0..<5 {
                let index = (head - trail + perimeter.count) % perimeter.count
                let (x, y) = perimeter[index]
                setPixel(&buf, x: x, y: y, color: processingRail, intensity: 1 - Double(trail) * 0.17)
            }
        case .awaiting:
            let phase = ((animFrame / 4) & 1) == 0
            let points = phase
                ? [(0, 0), (1, 0), (0, 1), (10, 10), (9, 10), (10, 9)]
                : [(10, 0), (9, 0), (10, 1), (0, 10), (1, 10), (0, 9)]
            for (x, y) in points { setPixel(&buf, x: x, y: y, color: awaitingRail) }
        case .error:
            let intensity = 0.65 + 0.35 * ((sin(Double(animFrame) * 0.35) + 1) / 2)
            for index in stride(from: 0, to: perimeter.count, by: 2) {
                let (x, y) = perimeter[index]
                setPixel(&buf, x: x, y: y, color: errorRail, intensity: intensity)
            }
        case .idle:
            let intensity = 0.56 + 0.16 * ((sin(Double(animFrame) * 0.12) + 1) / 2)
            for (x, y) in [(0, 0), (10, 0), (10, 10), (0, 10)] {
                setPixel(&buf, x: x, y: y, color: idleRail, intensity: intensity)
            }
        }
    }

    /// Paint a complete 11×11 Agent Beacon frame.
    static func paintBeacon(
        _ buf: inout [UInt8], creature: MicroCreature?, aggregate: MicroAggregate, animFrame: Int
    ) {
        guard buf.count == size * size * 3 else { return }
        for i in 0..<(size * size) {
            buf[i * 3] = background.0
            buf[i * 3 + 1] = background.1
            buf[i * 3 + 2] = background.2
        }
        if let creature { paintOfficialMark(&buf, creature: creature, aggregate: aggregate) }
        else { paintStandby(&buf, animFrame: animFrame) }
        paintStatusRail(&buf, aggregate: aggregate, animFrame: animFrame)
    }
}
#endif
